File: module/pkg/models/signal.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type SignalStatus string

const (
	SignalActive    SignalStatus = "active"    // 활성화됨 (참여 가능)
	SignalFull      SignalStatus = "full"      // 정원 마감
	SignalClosed    SignalStatus = "closed"    // 마감됨 (시간 지남)
	SignalCancelled SignalStatus = "cancelled" // 취소됨
	SignalCompleted SignalStatus = "completed" // 완료됨
)

type Signal struct {
	ID          uint         `json:"id" gorm:"primaryKey"`
	CreatorID   uint         `json:"creator_id" gorm:"not null"`
	Title       string       `json:"title" gorm:"size:100;not null"`
	Description string       `json:"description" gorm:"size:500"`
	Category    InterestCategory `json:"category" gorm:"not null"`
	
	// 위치 정보
	Latitude  float64 `json:"latitude" gorm:"not null"`
	Longitude float64 `json:"longitude" gorm:"not null"`
	Address   string  `json:"address" gorm:"size:200"`
	PlaceName string  `json:"place_name" gorm:"size:100"`
	
	// 시간 정보
	ScheduledAt time.Time `json:"scheduled_at" gorm:"not null"`
	ExpiresAt   time.Time `json:"expires_at" gorm:"not null"`
	
	// 인원 정보
	MaxParticipants     int `json:"max_participants" gorm:"not null"`
	CurrentParticipants int `json:"current_participants" gorm:"default:1"`
	MinAge              int `json:"min_age" gorm:"default:0"`
	MaxAge              int `json:"max_age" gorm:"default:100"`
	
	// 상태
	Status SignalStatus `json:"status" gorm:"default:'active'"`
	
	// 추가 설정
	AllowInstantJoin bool   `json:"allow_instant_join" gorm:"default:true"`
	RequireApproval  bool   `json:"require_approval" gorm:"default:false"`
	GenderPreference string `json:"gender_preference" gorm:"size:10"` // any, male, female
	
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	Creator      User                 `json:"creator,omitempty" gorm:"foreignKey:CreatorID"`
	Participants []SignalParticipant  `json:"participants,omitempty" gorm:"foreignKey:SignalID"`
	ChatRoom     *ChatRoom            `json:"chat_room,omitempty" gorm:"foreignKey:SignalID"`
}

type ParticipantStatus string

const (
	ParticipantPending   ParticipantStatus = "pending"   // 승인 대기
	ParticipantApproved  ParticipantStatus = "approved"  // 승인됨
	ParticipantRejected  ParticipantStatus = "rejected"  // 거절됨
	ParticipantLeft      ParticipantStatus = "left"      // 나감
	ParticipantNoShow    ParticipantStatus = "no_show"   // 노쇼
)

type SignalParticipant struct {
	ID       uint              `json:"id" gorm:"primaryKey"`
	SignalID uint              `json:"signal_id" gorm:"not null"`
	UserID   uint              `json:"user_id" gorm:"not null"`
	Status   ParticipantStatus `json:"status" gorm:"default:'pending'"`
	Message  string            `json:"message" gorm:"size:200"`
	JoinedAt *time.Time        `json:"joined_at"`
	LeftAt   *time.Time        `json:"left_at"`
	
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	Signal Signal `json:"-" gorm:"foreignKey:SignalID"`
	User   User   `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

// DTO 구조체들
type CreateSignalRequest struct {
	Title       string           `json:"title" binding:"required,min=5,max=100"`
	Description string           `json:"description" binding:"max=500"`
	Category    InterestCategory `json:"category" binding:"required"`
	
	Latitude  float64 `json:"latitude" binding:"required"`
	Longitude float64 `json:"longitude" binding:"required"`
	Address   string  `json:"address" binding:"max=200"`
	PlaceName string  `json:"place_name" binding:"max=100"`
	
	ScheduledAt time.Time `json:"scheduled_at" binding:"required"`
	
	MaxParticipants     int    `json:"max_participants" binding:"required,min=2,max=20"`
	MinAge              int    `json:"min_age" binding:"min=0,max=100"`
	MaxAge              int    `json:"max_age" binding:"min=0,max=100"`
	AllowInstantJoin    bool   `json:"allow_instant_join"`
	RequireApproval     bool   `json:"require_approval"`
	GenderPreference    string `json:"gender_preference" binding:"oneof=any male female"`
}

type JoinSignalRequest struct {
	Message string `json:"message" binding:"max=200"`
}

type SearchSignalRequest struct {
	Latitude  float64          `json:"latitude"`
	Longitude float64          `json:"longitude"`
	Radius    float64          `json:"radius" binding:"max=50000"` // 최대 50km
	Category  InterestCategory `json:"category"`
	StartTime *time.Time       `json:"start_time"`
	EndTime   *time.Time       `json:"end_time"`
	Page      int              `json:"page" binding:"min=1"`
	Limit     int              `json:"limit" binding:"min=1,max=50"`
}

type SignalWithDistance struct {
	Signal   `json:",inline"`
	Distance float64 `json:"distance"` // 미터 단위
}

File: ios/lib/features/signal/data/models/signal.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

part 'signal.g.dart';

@JsonSerializable()
class Signal extends Equatable {
  final int id;
  @JsonKey(name: 'creator_id')
  final int creatorId;
  final String title;
  final String description;
  final String category;
  final double latitude;
  final double longitude;
  final String address;
  @JsonKey(name: 'place_name')
  final String? placeName;
  @JsonKey(name: 'scheduled_at')
  final DateTime scheduledAt;
  @JsonKey(name: 'expires_at')
  final DateTime expiresAt;
  @JsonKey(name: 'max_participants')
  final int maxParticipants;
  @JsonKey(name: 'current_participants')
  final int currentParticipants;
  @JsonKey(name: 'min_age')
  final int? minAge;
  @JsonKey(name: 'max_age')
  final int? maxAge;
  @JsonKey(name: 'allow_instant_join')
  final bool allowInstantJoin;
  @JsonKey(name: 'require_approval')
  final bool requireApproval;
  @JsonKey(name: 'gender_preference')
  final String? genderPreference;
  final String status;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;
  final User? creator;
  final List<SignalParticipant>? participants;
  @JsonKey(name: 'chat_room')
  final ChatRoom? chatRoom;

  const Signal({
    required this.id,
    required this.creatorId,
    required this.title,
    required this.description,
    required this.category,
    required this.latitude,
    required this.longitude,
    required this.address,
    this.placeName,
    required this.scheduledAt,
    required this.expiresAt,
    required this.maxParticipants,
    required this.currentParticipants,
    this.minAge,
    this.maxAge,
    required this.allowInstantJoin,
    required this.requireApproval,
    this.genderPreference,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.creator,
    this.participants,
    this.chatRoom,
  });

  factory Signal.fromJson(Map<String, dynamic> json) => _$SignalFromJson(json);
  Map<String, dynamic> toJson() => _$SignalToJson(this);

  Signal copyWith({
    int? id,
    int? creatorId,
    String? title,
    String? description,
    String? category,
    double? latitude,
    double? longitude,
    String? address,
    String? placeName,
    DateTime? scheduledAt,
    DateTime? expiresAt,
    int? maxParticipants,
    int? currentParticipants,
    int? minAge,
    int? maxAge,
    bool? allowInstantJoin,
    bool? requireApproval,
    String? genderPreference,
    String? status,
    DateTime? createdAt,
    DateTime? updatedAt,
    User? creator,
    List<SignalParticipant>? participants,
    ChatRoom? chatRoom,
  }) {
    return Signal(
      id: id ?? this.id,
      creatorId: creatorId ?? this.creatorId,
      title: title ?? this.title,
      description: description ?? this.description,
      category: category ?? this.category,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      address: address ?? this.address,
      placeName: placeName ?? this.placeName,
      scheduledAt: scheduledAt ?? this.scheduledAt,
      expiresAt: expiresAt ?? this.expiresAt,
      maxParticipants: maxParticipants ?? this.maxParticipants,
      currentParticipants: currentParticipants ?? this.currentParticipants,
      minAge: minAge ?? this.minAge,
      maxAge: maxAge ?? this.maxAge,
      allowInstantJoin: allowInstantJoin ?? this.allowInstantJoin,
      requireApproval: requireApproval ?? this.requireApproval,
      genderPreference: genderPreference ?? this.genderPreference,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      creator: creator ?? this.creator,
      participants: participants ?? this.participants,
      chatRoom: chatRoom ?? this.chatRoom,
    );
  }

  @override
  List<Object?> get props => [
    id,
    creatorId,
    title,
    description,
    category,
    latitude,
    longitude,
    address,
    placeName,
    scheduledAt,
    expiresAt,
    maxParticipants,
    currentParticipants,
    minAge,
    maxAge,
    allowInstantJoin,
    requireApproval,
    genderPreference,
    status,
    createdAt,
    updatedAt,
    creator,
    participants,
    chatRoom,
  ];
}

@JsonSerializable()
class SignalWithDistance extends Equatable {
  final Signal signal;
  final double? distance;

  const SignalWithDistance({
    required this.signal,
    this.distance,
  });

  factory SignalWithDistance.fromJson(Map<String, dynamic> json) => _$SignalWithDistanceFromJson(json);
  Map<String, dynamic> toJson() => _$SignalWithDistanceToJson(this);

  @override
  List<Object?> get props => [signal, distance];
}

@JsonSerializable()
class User extends Equatable {
  final int id;
  final String email;
  final String? username;
  @JsonKey(name: 'is_active')
  final bool isActive;
  final UserProfile? profile;

  const User({
    required this.id,
    required this.email,
    this.username,
    required this.isActive,
    this.profile,
  });

  factory User.fromJson(Map<String, dynamic> json) => _$UserFromJson(json);
  Map<String, dynamic> toJson() => _$UserToJson(this);

  @override
  List<Object?> get props => [id, email, username, isActive, profile];
}

@JsonSerializable()
class UserProfile extends Equatable {
  @JsonKey(name: 'display_name')
  final String? displayName;
  final String? bio;
  @JsonKey(name: 'profile_image_url')
  final String? profileImageUrl;
  final int age;
  final String gender;
  @JsonKey(name: 'manner_score')
  final double mannerScore;
  @JsonKey(name: 'total_ratings')
  final int totalRatings;

  const UserProfile({
    this.displayName,
    this.bio,
    this.profileImageUrl,
    required this.age,
    required this.gender,
    required this.mannerScore,
    required this.totalRatings,
  });

  factory UserProfile.fromJson(Map<String, dynamic> json) => _$UserProfileFromJson(json);
  Map<String, dynamic> toJson() => _$UserProfileToJson(this);

  @override
  List<Object?> get props => [displayName, bio, profileImageUrl, age, gender, mannerScore, totalRatings];
}

@JsonSerializable()
class SignalParticipant extends Equatable {
  final int id;
  @JsonKey(name: 'signal_id')
  final int signalId;
  @JsonKey(name: 'user_id')
  final int userId;
  final String status;
  final String? message;
  @JsonKey(name: 'joined_at')
  final DateTime? joinedAt;
  @JsonKey(name: 'left_at')
  final DateTime? leftAt;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;
  final User? user;

  const SignalParticipant({
    required this.id,
    required this.signalId,
    required this.userId,
    required this.status,
    this.message,
    this.joinedAt,
    this.leftAt,
    required this.createdAt,
    required this.updatedAt,
    this.user,
  });

  factory SignalParticipant.fromJson(Map<String, dynamic> json) => _$SignalParticipantFromJson(json);
  Map<String, dynamic> toJson() => _$SignalParticipantToJson(this);

  @override
  List<Object?> get props => [id, signalId, userId, status, message, joinedAt, leftAt, createdAt, updatedAt, user];
}

@JsonSerializable()
class ChatRoom extends Equatable {
  final int id;
  @JsonKey(name: 'signal_id')
  final int signalId;
  final String name;
  @JsonKey(name: 'is_active')
  final bool isActive;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  const ChatRoom({
    required this.id,
    required this.signalId,
    required this.name,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ChatRoom.fromJson(Map<String, dynamic> json) => _$ChatRoomFromJson(json);
  Map<String, dynamic> toJson() => _$ChatRoomToJson(this);

  @override
  List<Object?> get props => [id, signalId, name, isActive, createdAt, updatedAt];
}

// 요청 모델들
@JsonSerializable()
class CreateSignalRequest extends Equatable {
  final String title;
  final String description;
  final String category;
  final double latitude;
  final double longitude;
  final String address;
  @JsonKey(name: 'place_name')
  final String? placeName;
  @JsonKey(name: 'scheduled_at')
  final DateTime scheduledAt;
  @JsonKey(name: 'max_participants')
  final int maxParticipants;
  @JsonKey(name: 'min_age')
  final int? minAge;
  @JsonKey(name: 'max_age')
  final int? maxAge;
  @JsonKey(name: 'allow_instant_join')
  final bool allowInstantJoin;
  @JsonKey(name: 'require_approval')
  final bool requireApproval;
  @JsonKey(name: 'gender_preference')
  final String? genderPreference;

  const CreateSignalRequest({
    required this.title,
    required this.description,
    required this.category,
    required this.latitude,
    required this.longitude,
    required this.address,
    this.placeName,
    required this.scheduledAt,
    required this.maxParticipants,
    this.minAge,
    this.maxAge,
    required this.allowInstantJoin,
    required this.requireApproval,
    this.genderPreference,
  });

  factory CreateSignalRequest.fromJson(Map<String, dynamic> json) => _$CreateSignalRequestFromJson(json);
  Map<String, dynamic> toJson() => _$CreateSignalRequestToJson(this);

  @override
  List<Object?> get props => [
    title,
    description,
    category,
    latitude,
    longitude,
    address,
    placeName,
    scheduledAt,
    maxParticipants,
    minAge,
    maxAge,
    allowInstantJoin,
    requireApproval,
    genderPreference,
  ];
}

@JsonSerializable()
class SearchSignalRequest extends Equatable {
  final double? latitude;
  final double? longitude;
  final double? radius;
  final String? category;
  @JsonKey(name: 'start_time')
  final DateTime? startTime;
  @JsonKey(name: 'end_time')
  final DateTime? endTime;
  final int page;
  final int limit;

  const SearchSignalRequest({
    this.latitude,
    this.longitude,
    this.radius,
    this.category,
    this.startTime,
    this.endTime,
    this.page = 1,
    this.limit = 20,
  });

  factory SearchSignalRequest.fromJson(Map<String, dynamic> json) => _$SearchSignalRequestFromJson(json);
  Map<String, dynamic> toJson() => _$SearchSignalRequestToJson(this);

  @override
  List<Object?> get props => [latitude, longitude, radius, category, startTime, endTime, page, limit];
}

@JsonSerializable()
class JoinSignalRequest extends Equatable {
  final String? message;

  const JoinSignalRequest({this.message});

  factory JoinSignalRequest.fromJson(Map<String, dynamic> json) => _$JoinSignalRequestFromJson(json);
  Map<String, dynamic> toJson() => _$JoinSignalRequestToJson(this);

  @override
  List<Object?> get props => [message];
}

// 상수 정의
class SignalStatus {
  static const String active = 'active';
  static const String full = 'full';
  static const String closed = 'closed';
  static const String cancelled = 'cancelled';
  static const String completed = 'completed';
}

class ParticipantStatus {
  static const String pending = 'pending';
  static const String approved = 'approved';
  static const String rejected = 'rejected';
  static const String left = 'left';
  static const String noShow = 'no_show';
}

class InterestCategory {
  static const String sports = 'sports';
  static const String food = 'food';
  static const String culture = 'culture';
  static const String study = 'study';
  static const String hobby = 'hobby';
  static const String travel = 'travel';
  static const String shopping = 'shopping';
  static const String entertainment = 'entertainment';
}